Fix rescan support for external scan nodes.

hlinnaka opened the pull request:
Instead of pushing the responsibility to rescanning down to each different
kind of external table, custom, EXECUTE, or file, implement rescanning
in fileam.c in a generic fashion, by closing and reopening the underlying
"url". This gets us rescan support for custom and EXECUTE-type external
tables, which was missing before. The missing support caused a crash if you
have a plan that requires rescanning, as the one added in the regression
test.

ivannovick commented:
@yydzero @vraghavan78 can you please review this PR?

yydzero commented:
@ivannovick according to discussion in another email thread, this is more a PM call from my point of view.

hlinnaka commented:
On 09/03/16 16:27, Yandong Yao wrote:
> @ivannovick according to discussion in another email thread, this is more a PM call from my point of view.

We need to fix rescan support in any case, even if we also teach the 
planner to add a Materialize node to avoid the rescanning. All executor 
nodes are expected to have working rescan support, and adding 
Materialize nodes would only be an optimization, not something we can 
rely on 100%. Even if we cover all the cases now, future planner changes 
might introduce new cases where a rescan happens.

- Heikki

danielgustafsson commented:
I think this is the cleanest way to reliably support rescans, which is something we have to do regardless like @hlinnaka says. If rescanning really becomes a performance problem can't we add a cost estimate along the line of `external_table_rescan_cost=x.y` to be able to influence the planner to weight against plans that favour rescanning?

atris commented:
IMO rescanning isnt really a cost based decision, it is a side effect of
other operators that proved to be most cost efficient and require
rescanning. Postgres assumes all operators are rescannable, so adding a
cost factor for that may alter the shape of the whole plan tree giving us
potentially less optimal plans..

On Thu, Mar 10, 2016 at 2:00 PM, Daniel Gustafsson <notifications@github.com
> wrote:

> I think this is the cleanest way to reliably support rescans, which is
> something we have to do regardless like @hlinnaka
> <https://github.com/hlinnaka> says. If rescanning really becomes a
> performance problem can't we add a cost estimate along the line of
> external_table_rescan_cost=x.y to be able to influence the planner to
> weight against plans that favour rescanning?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/490#issuecomment-194732376>.
>



-- 
Regards,

Atri
*l'apprenant*

hlinnaka commented:
Right. I admit I haven't looked at the code, but I would assume 
rescanning to be assign costed the same as the node's startup cost, 
which for external tables should be fairly significant already. That 
seems about right, I don't think fudging that further would be an 
improvement. With those costs, the planner should already choose a plan 
that doesn't involve rescanning an external table, if it knows how to 
generate such a plan at all.

On 10/03/16 10:33, Atri Sharma wrote:
> IMO rescanning isnt really a cost based decision, it is a side effect of
> other operators that proved to be most cost efficient and require
> rescanning. Postgres assumes all operators are rescannable, so adding a
> cost factor for that may alter the shape of the whole plan tree giving us
> potentially less optimal plans..
>
> On Thu, Mar 10, 2016 at 2:00 PM, Daniel Gustafsson <notifications@github.com
>> wrote:
>
>> I think this is the cleanest way to reliably support rescans, which is
>> something we have to do regardless like @hlinnaka
>> <https://github.com/hlinnaka> says. If rescanning really becomes a
>> performance problem can't we add a cost estimate along the line of
>> external_table_rescan_cost=x.y to be able to influence the planner to
>> weight against plans that favour rescanning?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/greenplum-db/gpdb/pull/490#issuecomment-194732376>.
>>
>
>
>


-- 

- Heikki

danielgustafsson commented:
Fair enough, those are good points.

hlinnaka commented:
@yydzero, @vraghavan78, can you have a look at this patch, please?

foyzur commented:
I love this fix!! That's what @vraghavan78 and I discussed previously. Whether to pick rescanning is a planner/optimizer decision. But, a scan node, implementing an abstract scanner interface (just for an example) shouldn't throw exception at runtime claiming that it doesn't have rescanning.

vraghavan78 commented:
@hlinnaka I agree with this PR. But I think @ivannovick has to address the product question.

@ivannovick: Can you please decide from the product perspective if this is an acceptable behavior?

ivannovick commented:
Hi guys,  

From user perspective, I don't see any reason not to proceed with this patch.  Which part of the behavior change is might be considered bad or non-obvious to users?

Cheers,
Ivan

vraghavan78 commented:
In src/include/nodes/relation.h we have the following statement:

```
	bool		isrescannable; /* false for ext web tables */
```

that needs to be fixed.

vraghavan78 commented:
The following function in *./src/backend/optimizer/util/plancat.c* needs to be fixed:

```
get_external_relation_info(Oid relationObjectId, RelOptInfo *rel)
{

...
	/* any external tables are non-rescannable. */
	rel->isrescannable = false;

}
```

vraghavan78 commented:
LGTM

yydzero commented:
LGTM from technical point of view.

@ivannovick the change is old behavior will error out when rescan external table, while new behavior will do the rescan, thus might result in different results between first scan and 2nd scan.

lij55 commented:
@adam8157 , could you please also review it and check the effect on S3?

adam8157 commented:
external_stopscan() -> url_fclose() -> InvokeExtProtocol() -> `extprotocol->prot_last_call = true`

S3 handles that, LGTM.

ivannovick commented:
+1, mostly based on Venky's and Heikki's +1.  If you have any strong objection please make it known

hlinnaka commented:
Removed the RelOptInfo.isrescannable field and added a comment in create_external_path() to explain why external tables are marked as "not rescannable". This addresses @vraghavan78's comments.

danielgustafsson commented:
:+1:

hlinnaka commented:
Closing this in favour of PR #663. See that PR for details.
